### PR TITLE
[pipeline] add native pipeline ubo uploading

### PR DESCRIPTION
**Pull request by star-e**
native-pipeline hello world. Supports basic cyberpunk demo.

### Changelog

* remove texture1d default resource from gfx. (some platform such as webgl does not support texture1d)
* added ubo uploading and descriptor set to native pipeline
* fixed cc_joints size error in custom pipeline
* remove input attachment from deferred-lighting shader. Mixing subpass shader with plain shader is not supported any more.

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

File: cocos/gfx/base/device.ts [hunk 1]
```diff
@@ -354,11 +355,10 @@ export abstract class Device {
 }
 
 export class DefaultResource {
-    private _texture2D : Texture | null = null;
-    private _texture3D : Texture | null = null;
-    private _textureCube : Texture | null = null;
-    private _texture1DArray : Texture | null = null;
```
Comment (star-e) [inline cocos/gfx/base/device.ts:360]: remove texture1d which is not supported on gl backends

File: editor/assets/effects/pipeline/deferred-lighting.effect [hunk 2]
```diff
@@ -44,36 +44,16 @@ CCProgram lighting-fs %{
...
-    layout(binding = 1) uniform sampler2D gbuffer_normalMap;
-    #pragma rate gbuffer_emissiveMap pass
-    layout(binding = 2) uniform sampler2D gbuffer_emissiveMap;
-    #pragma rate depth_stencil pass
-    layout(binding = 3) uniform sampler2D depth_stencil;
-  #endif
-
-  #if !CC_DEVICE_CAN_BENEFIT_FROM_INPUT_ATTACHMENT || __VERSION__ >= 450
-    layout(location = 0) out vec4 fragColor;
-  #endif
```
Comment (star-e) [inline editor/assets/effects/pipeline/deferred-lighting.effect:75]: We shall not mix subpass shader and plain shader

File: editor/assets/effects/pipeline/post-process.effect [hunk 3]
```diff
@@ -22,15 +22,15 @@ CCEffect %{
 
 CCProgram post-process-vs %{
   precision highp float;
-  #include <legacy/input-standard>
+  #include <legacy/decode-standard>
```
Comment (star-e) [inline editor/assets/effects/pipeline/post-process.effect:25]: decouple post process with skinning

File: native/cocos/core/assets/EffectAsset.h [hunk 4]
```diff
@@ -544,6 +544,12 @@ struct IShaderInfo {
         if (version == "glsl4") return &glsl4;
         return nullptr;
     }
+    IShaderSource *getSource(const ccstd::string &version) {
```
Comment (star-e) [inline native/cocos/core/assets/EffectAsset.h:547]: ~~add const version~~
add mutable versioin

File: native/cocos/renderer/gfx-base/GFXDevice.cpp [hunk 5]
```diff
@@ -158,14 +158,6 @@ DefaultResource::DefaultResource(Device *device) {
...
         device->copyBuffersToTexture(&bufferData, _texture3D, &region, 1);
     }
-    if (device->getCapabilities().maxArrayTextureLayers >= 2) {
-        _texture1DArray = device->createTexture({TextureType::TEX1D_ARRAY, TextureUsageBit::STORAGE | TextureUsageBit::SAMPLED,
-                                                 Format::RGBA8, 1, 1, TextureFlagBit::NONE, 2});
-        BufferTextureCopy region = {0, 0, 0, {0, 0, 0}, {1, 1, 1}, {0, 0, 1}};
-        device->copyBuffersToTexture(&bufferData, _texture1DArray, &region, 1);
-        region.texSubres.baseArrayLayer = 1;
-        device->copyBuffersToTexture(&bufferData, _texture1DArray, &region, 1);
-    }
```
Comment (star-e) [inline native/cocos/renderer/gfx-base/GFXDevice.cpp:168]: remove texture1d

File: native/cocos/renderer/pipeline/custom/FrameGraphDispatcher.cpp [hunk 6]
```diff
@@ -809,6 +809,26 @@ void buildBarriers(FrameGraphDispatcher &fgDispatcher) {
...
+                } else {
+                    ++i;
+                }
+            }
+        };
+        removeUnnecessaryBarriers(node.blockBarrier.frontBarriers);
+        removeUnnecessaryBarriers(node.blockBarrier.rearBarriers);
+    }
+
+    // generate gfx barrier
```
Comment (star-e) [inline native/cocos/renderer/pipeline/custom/FrameGraphDispatcher.cpp:831]: temporary fix for unnecessary barriers

**Comment by star-e** [comment]
@cocos-robot run test cases

File: native/cocos/core/assets/EffectAsset.h (hunk 4 above)
Comment (minggo) [reply native/cocos/core/assets/EffectAsset.h:547]: It is a non-const version. Will modify it outside?

Comment (star-e) [reply native/cocos/core/assets/EffectAsset.h:547]: Yes, NativeProgramLibrary will overwrite bindings of shader source.

**Review by minggo — DISMISSED** [review]

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by suqionglian** [comment]
@cocos-robot run test cases

**Comment by suqionglian** [comment]
@cocos-robot run test cases

**Review by pandamicro — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Pull request merged**